### PR TITLE
Free'ing address space on scope delete

### LINKED ISSUE #4157: ERROR: could not find IP range for mask ffff0000 during network create
During docker-compose testing, I got the following error:
```
ERROR: could not find IP range for mask ffff0000
```

This was after a battery of compatibility testing.

VCH installation command:
```
./vic-machine-linux create --target=192.168.218.207 --image-store=datastore1 --name=vic-docker --user=root --password=xxxx --compute-resource="/ha-datacenter/host/esxbox.localdomain/Resources" --bridge-network=vic-network --debug=0 --no-tls --volume-store=datastore1/test:default --force
```

Reinstalling the VCH fixes this.  Container logs attached.
[container-logs (1).zip](https://github.com/vmware/vic/files/822967/container-logs.1.zip)

### PR DESCRIPTION
**Pull request by hmahmood**
Fixes #4157

### CODE REVIEW

File: lib/portlayer/network/context.go [hunk 1]
```diff
@@ -1258,6 +1258,35 @@ func (c *Context) DeleteScope(ctx context.Context, name string) error {
...
+
+	for _, p := range s.spaces {
+		for _, i := range append(s.dns, s.gateway, allzeros, allones) {
+			if !ip.IsUnspecifiedIP(i) {
+				p.ReleaseIP4(i)
+			}
+		}
+
+		if p.Parent != nil {
+			p.Parent.ReleaseIP4Range(p)
```
Comment (emlin) [inline lib/portlayer/network/context.go]: here has a check for if Parent is nil:
https://github.com/vmware/vic/blob/master/lib/portlayer/network/ipam.go#L334
Does that create problem?

Comment (emlin) [inline lib/portlayer/network/context.go]: never mind

File: lib/portlayer/network/context.go [hunk 2]
```diff
@@ -1258,6 +1258,35 @@ func (c *Context) DeleteScope(ctx context.Context, name string) error {
...
+		}
+	}
+
+	var parentSpace *AddressSpace
+	if len(s.spaces) == 1 {
+		parentSpace = s.spaces[0]
```
Comment (emlin) [inline lib/portlayer/network/context.go]: just curious why don't get parent space here?

Comment (hmahmood) [inline lib/portlayer/network/context.go]: in the case where `len(s.spaces) == 1`, the only space in the slice is the parent. See this case here: https://github.com/vmware/vic/blob/master/lib/portlayer/network/context.go#L443

File: lib/portlayer/network/context.go [hunk 3]
```diff
@@ -1258,6 +1258,35 @@ func (c *Context) DeleteScope(ctx context.Context, name string) error {
 		return fmt.Errorf("%s has active endpoints", s.Name())
 	}
 
+	var allzeros, allones net.IP
```
Comment (anchal-agrawal) [inline lib/portlayer/network/context.go]: Nit: `allZeros`, `allOnes`.

File: lib/portlayer/network/context_test.go [hunk 4]
```diff
@@ -18,16 +18,20 @@ import (
...
 	"reflect"
 	"strings"
 	"testing"
+	"time"
 
 	log "github.com/Sirupsen/logrus"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/mock"
 
+	"math/rand"
```
Comment (anchal-agrawal) [inline lib/portlayer/network/context_test.go]: We should put this in the top import group.

File: tests/test-cases/Group1-Docker-Commands/1-18-Docker-Network-RM.robot [hunk 5]
```diff
@@ -69,3 +69,17 @@ Remove network with running container
     ${rc}  ${output}=  Run And Return Rc And Output  docker %{VCH-PARAMS} network ls
     Should Be Equal As Integers  ${rc}  0
     Should Not Contain  ${output}  test-network
+
+Add and remove network multiple times
```
Comment (mhagen-vmware) [inline tests/test-cases/Group1-Docker-Commands/1-18-Docker-Network-RM.robot]: Prefer naming this - Verify user can create and delete many networks

Comment (hmahmood) [reply tests/test-cases/Group1-Docker-Commands/1-18-Docker-Network-RM.robot]: Its creating and deleting the same network over and over again, not creating a new network each time.

**Pull request merged**